### PR TITLE
Add an experimental GPG trainer

### LINKED ISSUE #3472: [GPG][new trainer] Add support to new `GPG` method
### Method description

- **short description**: In this work, we revisit the traditional Policy Gradient (PG) mechanism and propose a minimalist RL approach termed Group Policy Gradient (GPG). Unlike conventional methods, GPG directly optimizes the original RL objective, thus obviating the need for surrogate loss functions. By eliminating the critic and reference models, avoiding KL divergence constraints, and addressing the advantage and gradient estimation bias, our approach significantly simplifies the training process compared to Group Relative Policy Optimization (GRPO). Our approach achieves superior performance without relying on auxiliary techniques or adjustments. Extensive experiments demonstrate that our method not only reduces computational costs but also consistently outperforms GRPO across various unimodal and multimodal tasks.
- **paper link**: https://arxiv.org/abs/2504.02546

### Open source status

- [x] The method implementation is available
- [x] The model weights are available
- [x] The training datasets are available

### Provide useful links for the implementation

- source-code link: https://github.com/AMAP-ML/GPG
- model weights link: 
    - https://huggingface.co/GD-ML/Qwen2.5-Math-7B-GPG
- authors: 
    - GitHub: @cxxgtxy @lerogo @undyingjoker
    - Scholar: [Xiangxiang Chu](https://scholar.google.com/citations?user=jn21pUsAAAAJ), [Hailang Huang](https://scholar.google.com/citations?user=X0o0Ib8AAAAJ), [Xiao Zhang](https://github.com/undyingjoker), [Fei Wei](https://scholar.google.com/citations?user=xqrPe6gAAAAJ), [Yong Wang](https://www.semanticscholar.org/author/Yong-Wang/1683878) (AMAP, Alibaba Group)

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
Reward groups with no relative signal still count toward the loss denominator and shrink the update. I added an experimental `GPGTrainer` that corrects this using the fraction of scorable completions in informative groups and checks generation alignment and incompatible settings. All trainer tests pass, covering missing rewards, rounding residuals, process ordering, and a tiny model generation run.
Closes #3472.

> [!NOTE]
> **Medium Risk**
> New experimental RL trainer that rescales GRPO loss and rejects many config combinations; mistakes could change training dynamics, but scope is limited to the new module with explicit validation and tests.
> 
> **Overview**
> Adds an **experimental GPG (Group Policy Gradient)** stack on top of `GRPOTrainer`: `trl.experimental.gpg` exports `GPGConfig` and `GPGTrainer`, with docs (`gpg.md`), a **Paper Index** entry, and broad unit tests.
> 
> `GPGConfig` mainly changes GRPO defaults for the paper setup (`beta=0`, `scale_rewards="none"`, `loss_type="grpo"`, `router_aux_loss_coef=0`) and adds **`bias_correction`** (default `True`). `GPGTrainer` requires `GPGConfig`, enforces that `steps_per_generation * num_iterations` divides `gradient_accumulation_steps`, and when bias correction is on it **rejects** configs that would break the rescaling (KL, LUSPO, token masking, Liger, entropy/MoE aux, etc.).
> 
> The new behavior is **eq. 7-style gradient-bias correction**: after scoring, it gathers advantages, marks informative groups from scorable advantage spread, caches the fraction of valid completion slots per train/eval mode, zeros masked advantages, and **divides the GRPO loss** by that fraction (skipped when the fraction is zero). `GPGTrainer` is also registered for telemetry in `base_trainer.py`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit baa58ada762940df6f71d93154ab7042c3ab1398. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/gpg/gpg_trainer.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        self._valid_group_fraction = 1.0
+
+    def _generate_and_score_completions(self, inputs):
+        inputs = super()._generate_and_score_completions(inputs)
+
+        # Read `frac_reward_zero_std` immediately after the parent appends it, rather than at loss time: the metric
+        # lists are cleared on every log, and a single generation batch feeds several gradient-accumulation
+        # micro-batches, each of which calls `_compute_loss`.
+        mode = "train" if self.model.training else "eval"
+        self._valid_group_fraction = 1.0 - self._metrics[mode]["frac_reward_zero_std"][-1]
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/gpg/gpg_trainer.py]: Confirmed and fixed in d8e39c3c. The finding is correct, and the mechanism is exactly as described.

Verified against the parent rather than reasoned about: `_prepare_inputs` buffers one train generation and reuses it for `steps_per_generation` optimizer steps, regenerating only every `steps_per_generation * num_iterations` steps (`grpo_trainer.py:1584-1592`), while evaluation generates for every batch (`:1596`). Both paths run through `_generate_and_score_completions`, so the single trainer-level scalar was written by whichever ran last. An eval landing inside a buffered train window left the remaining train micro-batches rescaling with the eval fraction.

The fix keys the cached fraction by mode, the way the parent already keys `_metrics`:

```python
self._valid_group_fraction = {"train": 1.0, "eval": 1.0}
...
self._valid_group_fraction[mode] = 1.0 - self._metrics[mode]["frac_reward_zero_std"][-1]
```

and `_compute_loss` reads back the fraction for its own mode.

Added `test_eval_does_not_clobber_the_train_correction_factor`, which sets the train fraction to 0.5, then an eval fraction of 1.0, and asserts a train-mode loss still divides by 0.5. Red before the fix, green after:

```
before d8e39c3c: FAILED test_eval_does_not_clobber_the_train_correction_factor
after  d8e39c3c: 5 passed
```

so the test pins the defect rather than passing vacuously.

One correction to the report, which does not change the conclusion: the trigger is not limited to `num_iterations > 1` or `steps_per_generation > gradient_accumulation_steps`. The buffer is consulted on every train step (`inputs = self._buffered_inputs[self._step % steps_per_generation]`), so any `steps_per_generation > 1` with an eval landing mid-window is enough.
